### DISCUSSION TITLE
Pin

### DISCUSSION THREAD

**Discussion by vmkmym** [KDT-villainlp/villainlp#1] [Announcements]
작업 진행하면서 겪었던 어려웠던 점들 기록해두기

**Comment by hongmyeoun** [comment]
이 밑으로 적는거지?

**Reply by vmkmym** [reply]
아니 그냥 내 메모였어...

**Reply by hongmyeoun** [reply]
+ troubleshoot 적어두기

**Comment by hongmyeoun** [comment]
어려웠던점(Comment)
offset이랑 그런 모디파이어 값들은 Boolean으로 처리하는데 
그 값을 처리할 때 ViewModel에서 처리하느냐 View에서 처리하느냐 고민
View에 관여하는 거라고 생각해 View에서만 처리했고, 
그 과정에서 Boolean값은 파라미터로 넘겨주면서 람다식에서 처리(최상위 함수에서만 처리)
-> 그러다보니 하위 함수들의 파라미터가 너무 중복되고 길어짐(문제발생)
-> 추후 최하위함수에서 변수를 생성해 해결하는 방향도 생각해보기

**Comment by hongmyeoun** [comment]
어려웠던점(Library)
조회수 업데이트와, 화면 불러오는 것이 가장 어려웠음, 
특히 코멘트의 수를 불러오는 함수는 따로 선언되어있어서 viewModel적용을 못함
Book이라는 data class로 선언된 객체값을 가져다가 사용했는데 
Firebase에서 DocumentId값으로 Book 객체 하나를 만들어내는 방식을 도입해야 될 것 같음
-> server package를 수정할 때 손보기

**Comment by hongmyeoun** [comment]
조회수 업데이트는 nav로 int값을 넘겨줬는데, 
getInt로 값을 받아올 수 없었음(추후 조사필요)
그래서 getString으로 받아오고 계산할 때 .toInt()를 사용
-> 그럴 일 없지만, 조회수 오버플로우가 났을때 문제발생

**Comment by vmkmym** [comment]
### chat/ground, openAichat, geminiChat

(문제) 기존 코드는 하나의 함수가 너무 일을 많이 했음. 
- 길게 엮어 놓은 것을 뷰모델을 새롭게 만들어가면서 리팩토링하는게 어려웠음.
- model, view, viewmodel을 분리하는 과정에서 비즈니스 로직과 view로직을 분리해야하는지 깨달음  
- 유지보수, 재사용성을 생각하면서 Refactoring 진행함
- boolean 변수는 전역변수로 선언해도 함수로 빼게 되면 빌드는 되지만 작동을 안함. 뷰모델로 시도해도 마찬가지. 람다식이 해결방법이었음. 

(문제) geminiChat을 적용해야함
- 창작마당(작가/꿈)에서 채팅리스트, 채팅방으로 이어지는 openAI 관련 로직을 완전히 분리하기 힘들었음.
- 그래서 openAI로직을 완전히 분리하는게 아니라 gemini를 필요한 곳에 선언해주는 방법으로 선택.
- 값을 넘겨줄 때 openAI와는 다른 로직으로 값을 전해주기 때문에 제대로 값 전달이 안되는 문제 발생.

**Comment by vmkmym** [comment]
### 전체 코드에서 문제점
- 하드코딩이 심하다
- object로 상수 선언하거나 enum class 활용할 생각을 못함
- 또는 컬러값, 문자열값 등으로 지정하면 되는데 전혀 안함
- 안좋은 개발 습관이므로 고쳐야됨
- 이번 리팩토링으로 뼈저리게 느끼는 중

**Comment by vmkmym** [comment]
## 드디어 해결한 문제 (ChattingList.kt)

```kotlin
.clickable {
                    if (novelInfo.threadId == "" || novelInfo.assistId == "") {
                        navController.navigate("GeminiChatScreen/${novelInfo.title}")
                    } else {
                        navController.navigate("ChattingScreen/${novelInfo.title}/${novelInfo.threadId}/${novelInfo.assistId}")
                    }
                }
``` 
1. novelInfo data class에서 threadId랑 assistId를 빈문자열로 선언해놓고 조건문에서 null이면 GeminiChatScreen으로 이동하게 함
그러니까 당연히 GeminiChatScreen으로 이동하지를 않는데 깨닫지 못하고 있었음.
2. novelInfo data class에서 threadId랑 assistId를 null값으로 하면 이동하려나? 싶었는데 생각해보니 조건문을 ""이렇게 선언해주면 될 것 같아서 해보니 됨.
3. 결론: 1개의 인자만 갖고 있으면 GeminiChatScreen, 3개의 인자를 갖고 있으면 ChattingScreen으로 이동하게 하며 문제 해결.

**Reply by vmkmym** [reply]
Realtime DB에 User가 보낸 메시지는 올라감!!

**Comment by hongmyeoun** [comment]
## 패키지 안에 파일이 있는데 import를 못하는 경우

1. Clean Project
2. Rebuild Project
3. File - invalidate Caches
